File: graphr/src/graphr/algorithms/DirectedSpreadingAgentPopulator.java
package graphr.algorithms;

import java.util.ArrayList;

import graphr.graph.Edge.Direction;
import graphr.processing.Agent;
import graphr.processing.AgentPopulator;

public class DirectedSpreadingAgentPopulator extends AgentPopulator {

	protected long queriedVertexId;
	protected long maxDistance;
	protected boolean startEverywhere = false;
	boolean settingVisibility = false;

	public DirectedSpreadingAgentPopulator(Long maxDistance){
		if(maxDistance!=null){
			this.maxDistance = maxDistance;
		}
		this.startEverywhere = true;
	}
	public DirectedSpreadingAgentPopulator(Long queriedVertexId, Long maxDistance) {
		if(queriedVertexId!=null){
			this.queriedVertexId = queriedVertexId;
		}
		if(maxDistance!=null){
			this.maxDistance = maxDistance;
		}
	}

	public DirectedSpreadingAgentPopulator(Long queriedVertexId, Long maxDistance, Boolean settingVisibility) {
		if(queriedVertexId!=null){
			this.queriedVertexId = queriedVertexId;
		}
		if(maxDistance!=null){
			this.maxDistance = maxDistance;
		}
		if(settingVisibility!=null){
			this.settingVisibility = settingVisibility;
		}
	}

	public ArrayList<Agent> getPopulation(long vertexId) {
		
		if (startEverywhere || vertexId == queriedVertexId ) {
			Agent a = createAgent(vertexId, maxDistance);
			ArrayList<Agent> list = new ArrayList<Agent>();
			list.add(a);
			return list;
		} else {
			return null;
		}
	}
	

	/**
	 * @param vertexId
	 *            The searched vertex id
	 * @param distance
	 *            The maximum distance
	 * @return a agent used to get the population
	 */
	protected Agent createAgent(final long vertexId, final long maxDistance) {
		return new DirectedSpreadingAgent(vertexId, maxDistance, settingVisibility);
	}

}

File: graphrs/graphrs/algorithms/DirectedSpreadingAgentPopulator.swift
//
//  DirectedSpreadingAgentPopulator.swift
//  graphrs
//
//  Created by Marcus Sellmann on 30.03.16.
//  Copyright © 2016 CITBDA. All rights reserved.
//

class DirectedSpreadingAgentPopulator: AgentPopulator {
    var queriedVertexId: UInt64 = 0
    var maxDistance: Int = 0
    var startEverywhere: Bool = false
    var settingVisibility: Bool = false
    
    init(maxDistance: Int) {
        self.maxDistance = maxDistance
        self.startEverywhere = true
    }
    
    init(queriedVertexId: UInt64, maxDistance: Int) {
        self.queriedVertexId = queriedVertexId
        self.maxDistance = maxDistance
    }
    
    init(queriedVertexId: UInt64, maxDistance: Int, settingVisibility: Bool) {
        self.queriedVertexId = queriedVertexId
        self.maxDistance = maxDistance
        self.settingVisibility = settingVisibility
    }
    
    func getPopulation(vertexId: UInt64, vertexProcessingFacade: VertexProcessingFacade) -> [Agent]? {
        if (startEverywhere || vertexId == queriedVertexId) {
            let a = createAgent(vertexId, maxDistance: maxDistance, vertexProcessingFacade: vertexProcessingFacade)
            return [a]
        } else {
            return nil
        }
    }
    
    /**
     * @param vertexId
     *            The searched vertex id
     * @param distance
     *            The maximum distance
     * @return a agent used to get the population
     */
    func createAgent(vertexId: UInt64, maxDistance: Int, vertexProcessingFacade: VertexProcessingFacade) -> Agent {
        return DirectedSpreadingAgent(sourceId: vertexId, maxDistance: maxDistance, setVisible: settingVisibility, vertexProcessingFacade: vertexProcessingFacade)
    }
}